src/v/cloud_storage/topic_recovery_service.cc

```diff
@@ -0,0 +1,599 @@
+          request.retention_bytes()};
+    }
+
+    if (request.retention_ms().has_value()) {
+        topic_properties.retention_local_target_ms
+          = tristate<std::chrono::milliseconds>{request.retention_ms()};
+        topic_properties.retention_duration
+          = tristate<std::chrono::milliseconds>{request.retention_ms()};
+    }
+
+    return topic_to_create_cfg;
+}
+
+ss::future<std::vector<cluster::topic_result>>
+topic_recovery_service::create_topics(const recovery_request& request) {
+    std::vector<cluster::topic_configuration> topic_configs;
+    topic_configs.reserve(_downloaded_manifests->size());
+
+    std::transform(
+      _downloaded_manifests->cbegin(),
+      _downloaded_manifests->cend(),
+      std::back_inserter(topic_configs),
+      [&request](const auto& m) { return make_topic_config(m, request); });
+
+    co_return co_await _topics_frontend.local().autocreate_topics(
+      topic_configs, config::shard_local_cfg().create_topic_timeout_ms());
+}
+
+ss::future<std::vector<cloud_storage::topic_manifest>>
+topic_recovery_service::filter_existing_topics(
+  const remote::list_bucket_items& items,
+  const recovery_request& request,
+  std::optional<model::ns> filter_ns) {
+    absl::flat_hash_map<ss::sstring, absl::flat_hash_set<ss::sstring>>
+      topic_index;
+
+    for (const auto& topic : _topic_state.local().all_topics()) {
+        topic_index.try_emplace(topic.ns, absl::flat_hash_set<ss::sstring>{});
+        topic_index[topic.ns].insert(topic.tp);
+    }
+
+    std::vector<topic_manifest> manifests;
+    manifests.reserve(items.size());
+
+    std::optional<std::regex> requested_pattern = std::nullopt;
+    if (request.topic_names_pattern().has_value()) {
+        requested_pattern.emplace(
+          request.topic_names_pattern().value().data(),
+          request.topic_names_pattern().value().size());
+    }
+
+    for (const auto& item : items) {
+        std::string s{item.key.data(), item.key.size()};
+
+        std::smatch matches;
+        const auto is_topic_manifest = std::regex_match(
+          s, matches, manifest_path_expr);
+        if (!is_topic_manifest) {
+            continue;
+        }
+
+        const auto& ns = matches[1].str();
+        const auto& tp = matches[2].str();
+
+        if (
+          requested_pattern.has_value()
+          && !std::regex_search(tp, requested_pattern.value())) {
+            vlog(
+              cst_log.debug,
+              "will skip topic {}, it does not match pattern {}",
+              tp,
+              request.topic_names_pattern().value());
+            continue;
+        }
+
+        if (topic_index.contains(ns) && topic_index[ns].contains(tp)) {
+            vlog(
+              cst_log.debug,
+              "will skip creating {}:{}, topic already exists",
+              ns,
+              tp);
+            continue;
+        }
+
+        if (auto download_r = co_await download_manifest(item.key);
```

Comment:
What happens if we fail to download some of the manifests? Reading the code, I think that we simply ignore that partition for the rest of the recovery process.

Reply:
Yes, any errors during download or parsing of a manifest result in that topic being skipped from recovery.